### ISSUE TITLE
New Feature: Work Orders

### ISSUE THREAD

**Issue by SteveIreland** [opensourcepos/opensourcepos#1539]
### New Feature / Enhancement

We've been using an enhancement I made to OSPOS for awhile now.  I recall submitting a pull request for it awhile back but pulled it because it was rushed and wasn't really well thought out in the context of a general purpose open source project.

I'm rewriting it so that it is much more flexible.  Essentially it will provide basic support for a small shop that needs to track repairs and services (in addition to providing general point of sales for parts and supplies).

Since the upcoming work on purchasing, receiving and vendor invoice reconciliation is on the near horizon I thought I'd repackage this and put some lipstick on it and submit another pull request for it to see if it can be incorporated in OSPOS.  The reason this is a prerequisite is that "work orders" will often generate job specific purchasing requirements so I will need to integrate with it and it will be much easier to integrate if it is part of the base.

To put it in context, when a customer approves a "quote" then the quote becomes a "work order".  When all of the work associated with the "work order" is complete then the work order becomes an "invoice".

**Comment by beinafuushop** [comment]
@SteveIreland Am not sure if i get how the work order functionality works, Can you please open the Wiki and put some insight for our understanding?

**Comment by SteveIreland** [comment]
Yes. The wiki will be updated when I submit the pull request this weekend.

**Comment by daN4cat** [comment]
I remember you submitted something about this as additional state to the sale transaction together with suspended, quote and etc. Right?
We should also consider the state cancel or delete. I know you don't fancy the soft delete but having numbered sale transitions it's better to do soft deletes in case of fiscal controls otherwise they will argue about the gaps in the enumeration sequence.

**Comment by SteveIreland** [comment]
Yes.  You are psychic.  I added support for it this last week. I had to add a canceled state to the sale_status field in order to handle a request that work orders not be lost when they are canceled.   They want to be able to generate a sales lost analysis report.  Your reason is more comprehensive so I'll go back and make sure that all cancellations are "soft".  (Auditors trump management every time).

**Comment by SteveIreland** [comment]
Documentation for Work Orders can now be found on the wiki at

[Work Orders](../wiki/Work-Orders)

Also once installed, records in the sales tables will no longer be physically deleted.  They will be "soft deleted".

**Comment by SteveIreland** [comment]
A pull request was submitted for Work Orders a few minutes ago.  It was based on the Office Menu feature which is now in a feature branch.

**Comment by odiea** [comment]
@SteveIreland Could you please check line 470 in database.sql .
Was
INSERT INTO `ospos_grants` (`permission_id`, `person_id`, 'menu_group') VALUES needs ` not ' around menu_group
worked this way
INSERT INTO `ospos_grants` (`permission_id`, `person_id`, `menu_group`) VALUES

**Comment by SteveIreland** [comment]
Will do.   Thanks.

**Comment by SteveIreland** [comment]
Fixed it.  The same issue is in the backoffice feature so I'll submit a fix for that now.

**Comment by odiea** [comment]
Are you gong to have a deleted filter like Items for Sales?

**Comment by SteveIreland** [comment]
I was wondering if anyone would think that would be useful.  If not then I was leaning toward creating a Reportico based report that would maybe group the deletions according to some cancellation reason convention which I hadn't  yet thought through.  Maybe by grouping on the comments and running a count of all cancellations with the same comment and a total on the dollar amounts of the sale.

**Comment by odiea** [comment]
Sorry To me that seems like overkill. Maybe just give the option to see deleted Sales In Detailed Reports with the option of undeleting them if necessary..

**Comment by odiea** [comment]
On the one I am playing with I removed the ability to delete Sales In takings and kept it in Detailed Sales Report.

**Comment by SteveIreland** [comment]
The guy funding my friend's startup asked to see an analysis of lost sales.  So that's why the Reportico approach.  However, since the filter on deleted sales with an option to "undelete" is something you feel will be useful we'll throw that in.  How hard can it be? :wink:

**Comment by odiea** [comment]
You could do it your way and just make Comments a mandatory field and if a reason is not placed in it then the delete button doesn't show. Then all you have to do is remove the delete button from the table view.

**Comment by SteveIreland** [comment]
On the sales reports I will add a filter for "Work Orders" to the sales type and also add a filter for "Sales Status" so that "Completed", "Suspended", and "Canceled" will be selectable options with "Completed" a the default value.

I think "All" for the sales type should only reflect Sales, Invoices, and Returns.  I don't believe that Quotes and Work Orders should be included in a Sales Report along with completed sales since they are essentially "suspended" sales.

**Comment by daN4cat** [comment]
In feature/backoffice branch

**Comment by SteveIreland** [comment]
Thanks for going ahead and merging it in.  I'll be following it this week with an improvement to the reporting side so that we have the ability to filter on work orders (and the filtering options are actually dependent on how the system is configured).  I also plan on starting the purchasing development

**Comment by daN4cat** [comment]
feature/backoffice is rebased with latest master

**Comment by SteveIreland** [comment]
Thanks. I should have the next wave of changes submitted tonight (after I've reviewed my language elements for conformance to the new rules of behavior)

**Comment by SteveIreland** [comment]
I tried to rebase on to the feature branch but there were errors that I just didn't understand so I spent the evening rebuilding a new branch that is based on the feature/branch.

I need a little more time to rebuild a test environment from the feature branch, apply my changes and make sure everything still works before submitting a pull request for it.

**Comment by daN4cat** [comment]
@SteveIreland don't know, but you'd better take the latest since I rebased again with latest master.

**Comment by SteveIreland** [comment]
ack ... ack ... ack

I was testing, debugging it and finalizing it and then submitted the pull request before I saw that  you moved my cheese.

**Comment by SteveIreland** [comment]
I'll go ahead and  catch it back up and resubmit it.

**Comment by SteveIreland** [comment]
I think I've rebased it correctly ...it's hard for me to tell. Since it has a big red X on it I have to assume something about it isn't right.  I'll just update the wiki and wait on you to tell me where the pull request stands.

**Comment by daN4cat** [comment]
What's left of this task outside the merging to master and the tasks I highlighted earlier?

**Comment by SteveIreland** [comment]
After I make those two changes this weekend and doing a bit more testing I plan on installing into production. So I think that should be the end of the work order story

**Comment by daN4cat** [comment]
Ok good and thanks again for the contribution

**Comment by daN4cat** [comment]
@SteveIreland all is in master now.

Please note that the database upgrade scripts are now in one called 3.1.1_to_3.2.0.sql as master is now 3.2.0 version (and WIP).

**Comment by SteveIreland** [comment]
Okay. Thanks!

**Comment by daN4cat** [comment]
@SteveIreland reports_ language is missing some entries, e.g. Sale Type -> reports_complete. The drop down shows (TBD)

**Comment by SteveIreland** [comment]
Those have been added and a pull request has been submitted for the changes.

**Comment by daN4cat** [comment]
I merged. However I have the following observations:

- I don't see a means to get deleted sales listed. Details sales doesn't show anything even because the drop down menu has Completed Sales & Returns, Completed Sales and Returns (are the Returns completed or no?)
- The delete button on the pop up dialog box should be removed if the user has no rights to cancel the sale

**Comment by SteveIreland** [comment]
When you display the drop-down list for Sales Type, canceled should be one of the options.  Are you not seeing this - or do I misunderstand the issue?

![image](https://user-images.githubusercontent.com/23066623/30250479-242277f0-961d-11e7-89c7-3038e2346c9b.png)

Also,I agree with you on the disabled options.  I mentioned that earlier and that the issue was that I was spending a lot of time trying to track down where all of that needs to take place.  The delete button is tightly coupled with some JavaScript magic that I'm not yet up to speed on.  So for now I just put in the controls that I was able to accomplish easily.

However, I expect that my depth of understanding will improve as I work on the purchasing module so as soon as I hit that "Ah-ha!" moment I will submit a change for that (unless someone wants to make the correction earlier).

**Comment by daN4cat** [comment]
@SteveIreland no I don't see Sale Type -> Cancelled.

**Comment by daN4cat** [comment]
I've checked the code and that option is only active if invoice and work order are active:

```
	public function get_sale_type_options()
	{
		$sale_type_options = array();
		$sale_type_options['complete'] = $this->lang->line('reports_complete');
		$sale_type_options['sales'] = $this->lang->line('reports_completed_sales');
		if($this->config->item('invoice_enable') == '1')
		{
			$sale_type_options['quotes'] = $this->lang->line('reports_quotes');
			if($this->config->item('work_order_enable') == '1')
			{
				$sale_type_options['work_orders'] = $this->lang->line('reports_work_orders');
				$sale_type_options['canceled'] = $this->lang->line('reports_canceled');
			}
		}
		$sale_type_options['returns'] = $this->lang->line('reports_returns');
		return $sale_type_options;
	}
```

I would say that needs to be valid in any case not only for invoices nor for work order

**Comment by SteveIreland** [comment]
Good catch.  I obviously have not done a thorough job of testing this in various scenarios.  It's all about my world you know.  :-)

I'll make that correction immediately after lunch!

**Comment by daN4cat** [comment]
Good. It should be trivial. Also are you considering a Takings filter option to display cancelled sales?

**Comment by SteveIreland** [comment]
A pull request for the missing Cancel option has been submitted.

**Comment by SteveIreland** [comment]
@daN4cat I wasn't planning on adding the filter there but that would make sense.  Should that be done now or can it wait until I figure out how to relabel the "Delete" button to "Suspend" when the "Canceled" filter is selected?

**Comment by daN4cat** [comment]
When you like, it was the extra suggestion. You may want to consider in general what else you want to add to those filters given the new feature.

**Comment by SteveIreland** [comment]
Will do.  I'm going to need to take the time to figure that all out as part of the purchasing development so this will definitely be revisited.

**Comment by beinafuushop** [comment]
Hello Team,

once you are done with development,please let us know and we will start
testing.

**Comment by SteveIreland** [comment]
I just noticed (in a production site) that I could not  update a sale.  It's not something they do very often so it's not a crisis, but I'm going to submit a pull request to fix this issue.  I was trying to update the comment on a canceled sale.

**Comment by SteveIreland** [comment]
I just submitted a pull request to fix the sale update issue.

**Comment by daN4cat** [comment]
@SteveIreland is anything left to close this feature?

**Comment by SteveIreland** [comment]
Please close.  Thank you

**Comment by beinafuushop** [comment]
sorry team please help me to understand the process, Once the feature is
closed,does it mean it ready for release? I.e. development,testing,fixing
of bugs and etc are all completed?

**Comment by daN4cat** [comment]
It means it's in master and considered completed. Of course further testing could spot some issue due to a corner case, but that happens always.

**Comment by beinafuushop** [comment]
thanks for clarification.
